test/binary_archives/test_binary_archives.py

```diff
@@ -0,0 +1,89 @@
+    metadata = load_json(orchestra.orchestra_root / "share/orchestra/component_A.json")
+    assert metadata["binary_archive_path"] == expected_binary_archive_path
+
+    binary_archive_full_path = orchestra.binary_archives_dir / "origin/linux-x86-64/" / expected_binary_archive_path
+
+    expected_files = (
+        dedent(
+            """
+    bin/
+    include/
+    lib
+    lib64/
+    lib64/include/
+    lib64/pkgconfig/
+    libexec/
+    share/
+    share/doc/
+    share/man/
+    share/orchestra/
+    some_file
+    usr/
+    usr/include/
+    usr/lib/
+    """
+        )
+        .strip()
+        .splitlines()
+    )
+    files = subprocess.check_output(["tar", "tf", binary_archive_full_path], encoding="utf-8").splitlines()
+    files.sort()
+    assert files == expected_files
+
+
+def test_remote_heads_cache_poisoning_works(orchestra: OrchestraShim):
+    """Checks that the mechanism for poisoning the remote HEADs cache works"""
+    fake_commit = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+    fake_branch = "master"
+    orchestra.configuration.remote_heads_cache._set_entry("component_C", fake_branch, fake_commit)
+    component = orchestra.configuration.components["component_C"]
+    assert component.commit() == fake_commit
+    assert component.branch() == fake_branch
+
+
+def test_component_hashes(orchestra: OrchestraShim):
+    """Checks that the component hashes are as expected"""
+    # Test repositories are initialized each time, so, their commit IDs are not fixed.
+    # Getting reproducible commit hashes is difficult and brittle,
+    # and committing entire repositories is also not what we want.
+    # So we lie to orchestra and fix the commit hash of component_C in the cache.
+    fake_commit = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+    fake_branch = "master"
+    orchestra.configuration.remote_heads_cache._set_entry("component_C", fake_branch, fake_commit)
+
+    component = orchestra.configuration.components["component_C"]
+
+    assert component.self_hash == "504ea8d302f54bdd2404a48ed17ea5f86d2dee13"
+    assert component.recursive_hash == "f4ba860f9898fc41d14ec8d5dccf3a0f982d0bb7"
+
+
+def test_build_hashes(orchestra: OrchestraShim):
+    """Checks that the build hashes are as expected"""
+    # Test repositories are initialized each time, so, their commit IDs are not fixed.
+    # Getting reproducible commit hashes is difficult and brittle,
+    # and committing entire repositories is also not what we want.
+    # So we lie to orchestra and fix the commit hash of component_C in the cache.
+    fake_commit = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+    orchestra.configuration.remote_heads_cache._set_entry("component_C", "master", fake_commit)
+
+    component = orchestra.configuration.components["component_C"]
+
+    build = component.default_build
+    assert build.build_hash == "6da2984f62a5ab487c677bce39141242edb53734"
```

Comment:
I understand you don't want to reimplement the hashing in the test suite, but this makes tests very fragile.
Ideally you should have something like `assert build.build_hash == hash({ predictable_json_object })`.

As it is, if we change anything in the input to the hash function, the only thing we can do is update the test reference result, but it's not really unit test, it's just regression testing.
While adding a field to `predictable_json_object` would be way way more semantically rich.

Reply:
Now the component hashing tests do essentially this, meaning the hash is computed from information that is verified against a predictable value, and pytest will be able to display a rich diff between expected and actual value, were they to differ.